spec/support/vendor_api/open_api_3_specification.rb

```diff
@@ -0,0 +1,44 @@
+class OpenApi3Specification
+  def initialize(open_api_3_spec)
+    @spec = open_api_3_spec
+  end
+
+  def as_json_schema(schema_name)
+    spec = @spec.dup # we do not want to mutate the original schema
+
+    # Pull up the schema that we want to validate against into the top-level,
+    # so that json-schema understands it.
+    spec['$schema'] = 'http://json-schema.org/draft-04/schema#'
+    spec['$ref'] = "#/components/schemas/#{schema_name}"
+
+    schemas = spec['components']['schemas']
+    transformed_schemas = schemas.reduce({}) do |new_schemas, (name, definition)|
+      new_schemas[name] = transform_openapi_schema_to_json_schema(definition)
+      new_schemas
+    end
```

Comment:
that's pretty complex IMO,  we use reduce with an explicit assignment to the accumulator (new_schema).
It took me a while to grasp it, it might be simpler with a plain and explicit `.each`.
The following reduct is even more complex, with an IF inside the `reduce`,
suggestion: could we filter externally? instead of the filter inside the `reduce`?

Reply:
the `transform_openapi_schema_to_json_schema`  has been refactored, with a simpler and more semantic approach.